presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestJoinEnumerator.java

```diff
@@ -122,6 +152,93 @@ public void testDoesNotCreateJoinWhenPartitionedOnCrossJoin()
         assertEquals(actual.getCost(), PlanCostEstimate.infinite());
     }
 
+    @Test
+    public void testJoinClauseAndFilterInference()
+    {
+        ImmutableMap.Builder<String, Type> builder = ImmutableMap.builder();
+        builder.put("a", BIGINT);
+        builder.put("b", BIGINT);
+        builder.put("c", BIGINT);
+        builder.put("d", BIGINT);
+        Map<String, Type> variableMap = builder.build();
+        VariableReferenceExpression a = variable("a", variableMap.get("a"));
+        VariableReferenceExpression b = variable("b", variableMap.get("b"));
+        VariableReferenceExpression c = variable("c", variableMap.get("c"));
+        VariableReferenceExpression d = variable("d", variableMap.get("d"));
+
+        SymbolAliases.Builder newAliases = SymbolAliases.builder();
+        newAliases.put("A", new SymbolReference("a"));
+        newAliases.put("B", new SymbolReference("b"));
+        newAliases.put("C", new SymbolReference("c"));
+        newAliases.put("D", new SymbolReference("d"));
+        SymbolAliases symbolAliases = newAliases.build();
+
+        // Simple join predicates on variable references
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a = b"), ImmutableSet.of(a), ImmutableSet.of(b, c), "A = B", null);
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a = b", "c = d"), ImmutableSet.of(a, c), ImmutableSet.of(b, d), "A = B AND C = D", null);
+        // Complex join predicate - All variables must be from one join side to have the predicate be an equi-join clause
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a = b + c"), ImmutableSet.of(a), ImmutableSet.of(b, c), "A = B + C", null);
+        // Left and right side designation can be switched
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a = b + c"), ImmutableSet.of(b, c), ImmutableSet.of(a), "A = B + C", null);
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a = b + c + 1"), ImmutableSet.of(a), ImmutableSet.of(b, c), "A = B + C + 1", null);
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a = b + c + 1"), ImmutableSet.of(b, c), ImmutableSet.of(a), "A = B + C + 1", null);
+        // If a predicate has a mix of variables from left & right sides, the predicate is treated as a filter
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a + b = c"), ImmutableSet.of(a), ImmutableSet.of(b, c), null, "A + B = C");
+        // Test with multiple equi-join conditions and filters
+        assertJoinCondition(symbolAliases, toRowExpressionList(variableMap, "a = ABS(b)", "a = ceil(b-c)", "b = c + 10"),
+                ImmutableSet.of(a), ImmutableSet.of(b, c), "A = abs(B) AND A = ceil(B-C)", "B = C + 10");
+    }
```

Comment:
Also in the implementation, should we take determinism of row expression into consideration, i.e. skip non-deterministic expressions?

Reply:
We do the determinism check for predicates that participate in ReorderJoins [here](https://github.com/prestodb/presto/blob/69d60ccecc2aee6aa6c35d8592cae732e03abb77/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java#L720-L724). So any predicate expressions passed in here (to this internal method) are guaranteed to be deterministic

Reply:
I see. Can you maybe add a test case for it? like random(a) = b etc.

Reply:
https://github.com/prestodb/presto/blob/84fd0a4ec2d0167d4d385df086543880d6d8103c/presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestReorderJoins.java#L369 covers this case